File: zig/src/tools/eval_provider_secrets.zig
//! eval-provider-secrets — provider error secret-scrubbing parity runner.

const std = @import("std");
const zeroclaw = @import("zeroclaw");
const provider_secrets = zeroclaw.providers;

const EvalError = error{InvalidScenario};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout);
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();

    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;
    const input = getString(parsed.value, "input") orelse return EvalError.InvalidScenario;

    if (std.mem.eql(u8, op, "scrub")) {
        const output = try provider_secrets.scrubSecretPatterns(allocator, input);
        defer allocator.free(output);
        try writeOutput(writer, op, output);
    } else if (std.mem.eql(u8, op, "sanitize")) {
        const output = try provider_secrets.sanitizeApiError(allocator, input);
        defer allocator.free(output);
        try writeOutput(writer, op, output);
    } else {
        return EvalError.InvalidScenario;
    }
}

fn writeOutput(writer: anytype, op: []const u8, output: []const u8) !void {
    try writer.writeAll("{\"op\":");
    try std.json.stringify(op, .{}, writer);
    try writer.writeAll(",\"result\":{\"output\":");
    try std.json.stringify(output, .{}, writer);
    try writer.writeAll("}}\n");
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .string) return null;
    return inner.string;
}

File: eval-tools/src/bin/eval-provider-secrets.rs
//! eval-provider-secrets — provider error secret-scrubbing parity runner.

use std::io::{self, Read, Write};

use anyhow::{Context, Result};
use serde_json::Value;
use zeroclaw_providers::{sanitize_api_error, scrub_secret_patterns};

fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for raw_line in input.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = required_string(&value, "op")?;
        run_op(&mut output, op, &value)?;
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

fn run_op(output: &mut Vec<u8>, op: &str, value: &Value) -> Result<()> {
    let input = required_string(value, "input")?;
    match op {
        "scrub" => write_result(output, op, scrub_secret_patterns(input)),
        "sanitize" => write_result(output, op, sanitize_api_error(input)),
        _ => anyhow::bail!("unknown provider_secrets op: {op}"),
    }
    Ok(())
}

fn required_string<'a>(value: &'a Value, key: &str) -> Result<&'a str> {
    value
        .get(key)
        .and_then(Value::as_str)
        .with_context(|| format!("{key} missing"))
}

fn write_result(output: &mut Vec<u8>, op: &str, sanitized: String) {
    let value = serde_json::json!({ "op": op, "result": { "output": sanitized } });
    serde_json::to_writer(&mut *output, &value).expect("serialize eval result");
    output.push(b'\n');
}
